src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -1007,7 +1007,8 @@ protected void read(Symbol sym, int attrLen) {
                         //- System.err.println(" # " + sym.type);
                         if (sym.kind == MTH && sym.type.getThrownTypes().isEmpty())
                             sym.type.asMethodType().thrown = thrown;
-                        if (sym.kind == MTH  && sym.name == names.init && sym.owner.isValueClass()) {
+                        // Map value class factory methods back to constructors for the benefit of earlier pipeline stages
```

Comment:
At source levels less than the one where the value class feature is supported, javac incorrectly downgrades a value class to an ordinary class. (See the method com.sun.tools.javac.jvm.ClassReader#adjustClassFlags which internalizes the flags from a class file's header). As a result isValueClass returns false and we don't restore the correct return type of the constructor. This leads to the present problem. The present fix works around by not relying on isValueClass - the only instance where a constructor may have a non-Void return is the case of value class factory, so the check is redundant. 

https://bugs.openjdk.java.net/browse/JDK-8282525 has been raised for a fuller fix. 

A nuance for the reviewers: Source level constructors are lowered into static factory methods late in the pipeline stage and are remapped into constructors when read from class files - this is so as to minimize the changes to the compiler. Rather than know and deal with an entity called static value factory methods, most of the compiler simply works with constructors.

Reply:
Thanks for the extra explanation!!!